index.js

```diff
@@ -2,81 +2,113 @@
+ */
 if (!process.env["npm_config_argv"]) {
   console.log("This is meant to be run from within npm script. See https://github.com/charlesguse/run-script-os");
   return;
 }
 
+/**
+ * Executor process to match the script without blocking
+ */
 const spawn = require("child_process").spawn;
 
-// Switch to linux platform if cygwin/gitbash detected (fixes #7)
-// Allow overriding this behavior (fixes #11)
+/**
+ * Switch to linux platform if cygwin/gitbash detected (fixes #7)
+ * Allow overriding this behavior (fixes #11)
+ */
 let platform = process.platform;
 if (process.env.RUN_OS_WINBASH_IS_LINUX) {
   let shell = process.env.SHELL || process.env.TERM;
   shell = shell && shell.match("bash.exe") ? "bash.exe" : shell;
   platform = shell && ["bash.exe", "cygwin"].includes(shell) ? "linux" : process.platform;
 }
 
+/**
+ * Scripts as found on the user's package.json
+ */
 const scripts = require(path.join(process.cwd(), "package.json")).scripts;
 
+/**
+ * The script being executed can come from either lifecycle events or command arguments
+ */
 let npmArgs = JSON.parse(process.env["npm_config_argv"]);
 let options = npmArgs.original;
-if (!(options[0] === "run" || options[0] === "run-script")) {
-  options.unshift("run");
-}
 
-// Check for yarn without install command; fixes #13
-if (process.env.npm_config_user_agent.includes('yarn') && !options[1]) options[1] = 'install';
+let argument = options[1];
+let event = process.env["npm_lifecycle_event"];
 
-let osCommand = `${options[1]}:${platform}`;
-let foundMatch = true;
-
-if (!(osCommand in scripts)) {
-  foundMatch = false;
-  let regex = new RegExp(`^(${options[1]}):([a-zA-Z0-9-]*:)*(${platform})(:[a-zA-Z0-9-]*)*$`, "g");
-  for (let command in scripts) {
-    if (command.match(regex)) {
-      osCommand = command;
-      foundMatch = true;
-      break;
-    }
-  }
+/**
+ * Yarn support
+ * Check for yarn without install command; fixes #13
+ */
+if (process.env.npm_config_user_agent.includes('yarn') && !argument) {
+  argument = 'install';
 }
 
-// Check for :windows alias; fixes #3
-if (!foundMatch && (`${options[1]}:windows` in scripts) && platform === 'win32') {
-  osCommand = `${options[1]}:windows`;
-  foundMatch = true;
-}
+/**
+ * Prepare command structure to test the available scripts defined in "scripts: { ... }"
+ */
+let osCommand = `${options[1]}:${platform}`;
+
+/**
+ * Basic match upon the property name
+ * determine wether the name of the script exists in the list of defined scripts
+ */
+let foundMatch = osCommand in scripts;
 
-// Check for :macos alias; fixes #3
-if (!foundMatch && (`${options[1]}:macos` in scripts) && platform === 'darwin') {
-  osCommand = `${options[1]}:macos`;
-  foundMatch = true;
+if (!foundMatch) {
+  /**
+   * More in-depth match
+   * Execute the regular expression to help identify missing scripts
+   * It also tests for different aliases
+   */
+  osCommand = matchScript(argument, platform, scripts) || matchScript(event, platform, scripts);
```

Comment:
In the case of lifecycle events, npm_config_argv.original will always be the arguments of the entry point. So for instance when running npm version patch, the version script will be executed with npm_config_argv.original = ['version', 'patch'] and npm_lifecycle_event = 'version' then the postversion script will be executed with npm_config_argv.original = ['version', 'patch'] and npm_lifecycle_event = 'postversion'. In the second case, it will ret to re-run the existing version script during postversion because npm_config_argv.original takes presidence here. I believe this should instead be matchScript(event, platform, scripts) || matchScript(argument, platform, scripts); which still should be fine for simple npm run use cases because npm_lifecycle_event is still set to the original run script name in those cases.